### ISSUE TITLE
Add `dns.resolveTxt()`

### ISSUE THREAD

**Issue by ryanramage** [holepunchto/bare-dns#1]
https://nodejs.org/api/dns.html#dnsresolvetxthostname-callback

I need some code to get txt records in bare. It would be great to have this.

**Comment by kasperisager** [comment]
This requires pulling in https://c-ares.org so it's not an unsubstantial investment. That's not a "no" by the way, but rather a "we'll see what the future holds" 😄

**Comment by kasperisager** [comment]
@yassernasc Something for a rainy day perhaps!

**Comment by ryanramage** [comment]
yeah this one looks challenging. I tried to AI code it, but it choked and then recommended the c-ares because that is what node core does. Trying to store service public keys in txt records. I think for now I will write a module that is not a core thing, that is allowed to fallback to nslookup, host, of dig, if available. Maybe we will see what the future holds here, otherwise

**Comment by RangerMauve** [comment]
@ryanramage Did you end up publishing that module? I need DNS TXT records in order to support mapping DNS to hyperdrives. One alternative I've considered is bundling a DNS-over-HTTPS module and parsing records from there.

**Comment by ryanramage** [comment]
No, it became too complicated too quickly, and I dropped the need for bare support, so I could move past this. But yes, I think there could be a lot of neat things unlocked if this was added to the core bare dns module

**Comment by RangerMauve** [comment]
I'm using DoH for now. Here's what the code looks like for anyone coming to this in the future:

```JavaScript
  // TODO: Cache for offline use
  async resolveDNSToKey (hostname) {
    if (!globalThis.fetch) {
      globalThis.fetch = (await import('bare-fetch')).default
    }
    const subdomained = `_dnslink.${hostname}`

    const url = `${this.dnsResolver}?name=${subdomained}&type=TXT`

    const response = await fetch(url, {
      headers: { accept: 'application/dns-json' }
    })

    if (!response.ok) {
      throw new Error(`Unable to resolve DoH for ${hostname} ${await response.text()}`)
    }

    const { Answer: answers } = await response.json()

    for (let { name, data } of answers) {
      if (name !== subdomained || !data) {
        continue
      }
      if (data.startsWith('"')) {
        data = data.slice(1, -1)
      }
      if (!data.startsWith(DNSLINK_PREFIX)) {
        continue
      }
      return data.slice(DNSLINK_PREFIX.length)
    }

    throw new Error(`DNS-Link Record not found for TXT ${subdomained}`)
  }

```

You'll likely want to do your own caching for offline usage.

You can find a list of public resolvers here: https://dnsprivacy.org/public_resolvers/#dns-over-https-doh

**Comment by kasperisager** [comment]
In case anyone wants to take a stab at it, here's a branch that adds c-ares as a dependency: https://github.com/holepunchto/bare-dns/tree/cares. The JavaScript layer should expose a `DNSResolver` class (https://nodejs.org/api/dns.html#class-dnsresolver) that binds to `ares_channel_t` and uses https://docs.libuv.org/en/v1.x/poll.html for I/O polling. The global APIs like `dns.resolveTxt()` should then maintain a global resolver instance that is started and stopped as necessary.

**Comment by yassernasc** [comment]
I have an initial draft working fine, c-ares plus "uv_poll", PR coming soon. After that, It'll be far simpler to support more Record Types other than TXT.

**Comment by RangerMauve** [comment]
Great to hear, ty! I'll give it a shot this summer. 🥳
